src/interface/misc.rs

```diff
@@ -60,6 +64,27 @@ pub fn check_cageid(cageid: u64) {
     }
 }
 
+pub fn concat_iovec_to_slice(iovec: *const interface::IovecStruct, iovcnt: i32) -> Vec<u8> {
+    // ensure that the iovec pointer is valid and points to a valid iovec array
+    let iovecs = unsafe { slice::from_raw_parts(iovec, iovcnt as usize) };
+    let mut data = Vec::new();
+    for iovec in iovecs {
+        let slice = unsafe { slice::from_raw_parts(iovec.iov_base as *const u8, iovec.iov_len) };
+        data.extend_from_slice(slice);
+    }
+    data
+}
+
+pub fn log_from_slice(data: &[u8]) -> i32 {
+    match str::from_utf8(data) {
+        Ok(s) => {
+            log_to_stdout(s);
```

Comment:
log_to_stdout writes to the standard out stream regardless of the provided file descriptor, and this helper function is the only function that performs any sort of writing into a stream. what if the provided stream is standard error or any other stream?

Reply:
updated with stderr and for other streams